### DISCUSSION TITLE
Issues rendering sprites and 3D models in 0.6 at the same time

### DISCUSSION THREAD

**Discussion by cdsupina** [bevyengine/bevy#3891] [Q&A]
Hello everyone! I'm trying to migrate my [game](https://github.com/thetawavegame/thetawave) from Bevy 0.5 to 0.6. The core of my game takes place in 2D, but I had a 3D background that gave my game a distinctive look. In Bevy 0.5 I was using a perspective camera and didn't have any issues rendering the 2D sprites and the background 3D models. Below is the Bevy 0.5 code for the perspective camera I was using.

```rust
// Setup perspective camera in Bevy 0.5
commands.spawn_bundle(PerspectiveCameraBundle {
    transform: Transform::from_xyz(0.0, 0.0, 950.0)
        .looking_at(Vec3::ZERO, Vec3::Y),
    perspective_projection: PerspectiveProjection {
        far: 10000.0,
        ..Default::default()
    },
    ..Default::default()
});
```

Snapshot of game from Bevy 0.5:
![Capture](https://user-images.githubusercontent.com/15306815/152926560-b2cd6382-00f2-400c-a5d9-375f67d99625.PNG)

Since migrating to 0.6. I haven't been able to setup the camera in a way that lets me view the background 3D models and the sprites. Below is the perspective camera code I was using in Bevy 0.6 after I initially migrated (same as before).

```rust
// Setup perspective camera in Bevy 0.6
commands.spawn_bundle(PerspectiveCameraBundle {
    transform: Transform::from_xyz(0.0, 0.0, 950.0)
        .looking_at(Vec3::ZERO, Vec3::Y),
    perspective_projection: PerspectiveProjection {
        far: 10000.0,
        ..Default::default()
    },
    ..Default::default()
});
```

The image below is the game in Bevy 0.6. 3D models are visible but sprites are not appearing.

![Capture](https://user-images.githubusercontent.com/15306815/152923385-74a99baf-a4c5-4ff8-96ec-6e0e30d0c0ac.PNG)

The next thing I tried was a 2D orthographic camera using the following code:

```rust
let mut camera = OrthographicCameraBundle::new_2d();
camera.transform = Transform::from_xyz(0.0, 0.0, 1000.0);
camera.orthographic_projection.far = 10000.0;
```

Now the 2D sprites are appearing but not the background models.
![Capture](https://user-images.githubusercontent.com/15306815/152926358-1d9139a5-a42f-40d0-a1dc-fee44b754ff2.PNG)

I was wondering if anyone had insight into how to render both the sprites and the background models like I had from Bevy 0.5 with the perspective camera. I also tried playing around with the other camera properties but didn't have much luck. Thank you!

**Comment by cart** [comment]
This is probably partially because 3d renders _after_ 2d in 0.6:

https://github.com/bevyengine/bevy/blob/main/crates/bevy_core_pipeline/src/main_pass_driver.rs

However switching the order _does not_ resolve the issue:

```rust
impl Node for MainPassDriverNode {
    fn run(
        &self,
        graph: &mut RenderGraphContext,
        _render_context: &mut RenderContext,
        world: &World,
    ) -> Result<(), NodeRunError> {
        let extracted_cameras = world.get_resource::<ExtractedCameraNames>().unwrap();
        
        if let Some(camera_3d) = extracted_cameras.entities.get(CameraPlugin::CAMERA_3D) {
            graph.run_sub_graph(
                crate::draw_3d_graph::NAME,
                vec![SlotValue::Entity(*camera_3d)],
            )?;
        }

        if let Some(camera_2d) = extracted_cameras.entities.get(CameraPlugin::CAMERA_2D) {
            graph.run_sub_graph(
                crate::draw_2d_graph::NAME,
                vec![SlotValue::Entity(*camera_2d)],
            )?;
        }
        Ok(())
    }
}
```

We should definitely do that, as 2d-on-top-of-3d is a way more common scenario.  But this will require some investigation to resolve. I'm converting this discussion to an issue so we can track it better. Thanks for the report!

**Reply by cart** [reply]
#3902
